File: algorithims/leetcode/insertIntoBST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} val
 * @return {TreeNode}
 */
 var insertIntoBST = function(root, val) {
    let node = new TreeNode(val, null, null);
    console.log("Node:", node)
    if(root == null){
        root = node;
    } else {
        insertNode(root, node)
    }
    return root
};
var insertNode = function(root, node) {
    if (root.val < node.val){
        if(root.right === null){
            console.log(true)
            root.right = node
            
        } else {
            console.log("right recursive")
            insertNode(root.right, node);
        }
    } else {
        if(root.left === null){
            console.log(true)
            console.log(root)
            root.left = node
            console.log(root.left)
            
        } else {
            console.log("left recursive")
            insertNode(root.left, node)
        }
    }
}

File: algorithims/leetcode/InsertIntoBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class InsertIntoBST {
    public static void main(String[] args){
        TreeNode orignalBST = new TreeNode(4, new TreeNode(3),  new TreeNode(6));
        insertIntoBST(orignalBST, 2);
        System.out.println(orignalBST.toString());
    }
    public static TreeNode insertIntoBST(TreeNode root, int val) {
        TreeNode node = new TreeNode(val);
        if(root == null){
            root = node;
        } else {
            insertNode(root, node);
        }
        return root;
    }
    public static void insertNode(TreeNode root, TreeNode node){
        if(root.val < node.val){
            if(root.right == null){
                root.right = node;
            } else {
                insertNode(root.right, node);
            }
        } else {
            if(root.left == null){
                root.left = node;
            } else {
                insertNode(root.left, node);
            }
        }
    }
    public static class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode() {}
      TreeNode(int val) { this.val = val; }
      TreeNode(int val, TreeNode left, TreeNode right) {
          this.val = val;
          this.left = left;
         this.right = right;
      }
    }
    
}